### PR TITLE
Improve readability of ciphers/mixed_keyword_cypher.py

### PR DESCRIPTION
**Pull request by yanvoi**
fixes: #8622 
### Describe your change:
I've tried my best with refactoring the code - it still might be too little but that's how much I've managed to do with the time I had.

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by yanvoi** [comment]
well then I guess the original code doesn't pass the tests either

File: ciphers/mixed_keyword_cypher.py
Comment (cclauss) [inline ciphers/mixed_keyword_cypher.py:30]: Please make this a `set`.

File: ciphers/mixed_keyword_cypher.py
Comment (cclauss) [inline ciphers/mixed_keyword_cypher.py:43]: Please do all this with `set`

**Comment by cclauss** [comment]
On your local machine, please run `python3 -m doctest -v ciphers/mixed_keyword_cypher.py`
because pytest is failing on:
```
=================================== FAILURES ===================================
_____________ [doctest] ciphers.mixed_keyword_cypher.mixed_keyword _____________
008     H E L O
009     A B C D
010     F G I J
011     K M N P
012     Q R S T
013     U V W X
014     Y Z
015     and map vertically
016 
017     >>> mixed_keyword("college", "UNIVERSITY")  # doctest: +NORMALIZE_WHITESPACE
Expected:
    {'A': 'C', 'B': 'A', 'C': 'I', 'D': 'P', 'E': 'U', 'F': 'Z', 'G': 'O', 'H': 'B',
     'I': 'J', 'J': 'Q', 'K': 'V', 'L': 'L', 'M': 'D', 'N': 'K', 'O': 'R', 'P': 'W',
     'Q': 'E', 'R': 'F', 'S': 'M', 'T': 'S', 'U': 'X', 'V': 'G', 'W': 'H', 'X': 'N',
     'Y': 'T', 'Z': 'Y'}
    'XKJGUFMJST'
Got:
    'XKJGUFMJST'
```

File: ciphers/mixed_keyword_cypher.py
Comment (yanvoi) [reply ciphers/mixed_keyword_cypher.py:30]: It cannot be done with sets due to the nature of the algorithm. The order of those unique characters matters.

Comment (yanvoi) [reply ciphers/mixed_keyword_cypher.py:30]: I believe it cannot be done with sets because the order of these characters in the keyword matters - it determines how we will map plaintext characters to the cyphertext.

**Comment by yanvoi** [comment]
My PR failing the test was indeed my fault, I took the liberty of getting rid of print() from the body of the function which caused it to fail the test. My bad. Although I still believe that the printing of the mapping dictionary shouldn't be included in the code.

File: ciphers/mixed_keyword_cypher.py
Comment (yanvoi) [reply ciphers/mixed_keyword_cypher.py:43]: As this is a basic substitution cypher, it is bounded by the number of different letters in the alphabet we're using. Using sets in this case wouldn't change the performance almost at all.

**Comment by yanvoi** [comment]
It turned out that whoever has written this algorithm hard-coded the number of rows into it.

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

File: ciphers/mixed_keyword_cypher.py
Comment (yanvoi) [reply ciphers/mixed_keyword_cypher.py:27]: I can of course change that but I wanted to make it easier for people who would like to use other alphabets (e.g. cyrillic) for their ciphers. So it is your call - should I change it or not?

Comment (tianyizheng02) [reply ciphers/mixed_keyword_cypher.py:27]: If people want to use other alphabets such as Cyrillic, then they would have to go into the source code and change `ascii_uppercase` to a different variable. That doesn't really address the issue that `ascii_uppercase` (or whatever the user replaces it with) still doesn't need to be converted to a list.

If you want it to be easier to switch to a different alphabet, then perhaps you could have the function take in the alphabet string (`ascii_uppercase`, etc) as an input parameter

Comment (yanvoi) [reply ciphers/mixed_keyword_cypher.py]: Fair point - I've done both - Now there's no conversion to the list and a different alphabet can be passed to the function. Thank you for the review!

**Review by tianyizheng02 — APPROVED** [review]

**Comment by tianyizheng02** [comment]
> The tests seem to be failing due to code outside of the file I've been editing.

Other contributors have been experiencing the same issue with ruff... try running `ruff check --fix .` to see if ruff can automatically fix any of the linting issues

**Comment by yanvoi** [comment]
Sadly it resulted in even more, worse issues

**Comment by tianyizheng02** [comment]
@cclauss How is it that @rohan472000's changes in their own branch are making their way into this PR?

**Comment by tianyizheng02** [comment]
Oh my goddddd I knew this was gonna happen...

@yanvoi The ruff error is once again in another file that you didn't edit, but I already fixed it in #8743. Once that PR gets merged, updating your branch on GitHub should finally fix the issue

**Comment by yanvoi** [comment]
@tianyizheng02 No worries, I'll update when #8743 gets merged

**Comment by tianyizheng02** [comment]
@yanvoi The ruff error was fixed in #8746 instead, so you should be good to update your branch

**Comment by yanvoi** [comment]
@tianyizheng02 All done, thanks again!

File: ciphers/mixed_keyword_cypher.py
Comment (tianyizheng02) [reply ciphers/mixed_keyword_cypher.py]: > I believe it cannot be done with sets because the order of these characters in the keyword matters - it determines how we will map plaintext characters to the cyphertext.

Then I think it would help if this was mentioned in the comments, so that future maintainers know that the char ordeer matters

File: ciphers/mixed_keyword_cypher.py
Comment (tianyizheng02) [inline ciphers/mixed_keyword_cypher.py:61]: Ideally the algorithm shouldn't have side effects such as printing, but I understand that printing the mapping is helpful for debugging purposes. Perhaps it'd be better to have a "verbose" function parameter that allows users to enable/disable printing the mapping

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

File: ciphers/mixed_keyword_cypher.py
Comment (yanvoi) [reply ciphers/mixed_keyword_cypher.py]: Done - I've also added a test case that includes setting the verbose parameter to False

File: ciphers/mixed_keyword_cypher.py
Comment (tianyizheng02) [inline ciphers/mixed_keyword_cypher.py:5]:
```suggestion
    keyword: str, plaintext: str, verbose: bool = False, alphabet: str = ascii_uppercase
```

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

**Review by tianyizheng02 — APPROVED** [review]

**Review by cclauss — APPROVED** [review]

**Comment by Starmania** [comment]
Thanks you for the edit yanvoi !

**Pull request merged**
